### PR TITLE
[SILOptimizer] Add ‘llvm.cttz.*’ Intrinsic support to Constant Folding

### PR DESCRIPTION
**Pull request by shajrawi**
rdar://problem/29522851

### CODE REVIEW

**Comment by shajrawi** [comment]
@swift-ci Please test and merge

File: lib/SILOptimizer/Utils/ConstantFolding.cpp [hunk 1]
```diff
@@ -278,6 +278,44 @@ constantFoldBinaryWithOverflow(BuiltinInst *BI, BuiltinValueKind ID,
...
+  APInt LHSI = LHS->getValue();
+  unsigned LZ = 0;
+  // Check corner-case of source == zero
+  if (LHSI == 0) {
+    auto *RHS = dyn_cast<IntegerLiteralInst>(Args[1]);
+    if (!RHS || RHS->getValue() != 0) {
+      // Undefined
+      return nullptr;
+    }
+    LZ = LHSI.getBitWidth();
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/ConstantFolding.cpp]: Can you do an early exit here?

File: lib/SILOptimizer/Utils/ConstantFolding.cpp [hunk 2]
```diff
@@ -278,6 +278,44 @@ constantFoldBinaryWithOverflow(BuiltinInst *BI, BuiltinValueKind ID,
...
+  } else {
+    switch (ID) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/ConstantFolding.cpp]: My suggestion: Use an inline constructor:

unsigned LZ = [&] {
   ...
}();

Also. Do you really think we are going to support more types of intrinsics here? Why not just put in an assert or something like that.

File: lib/SILOptimizer/Utils/ConstantFolding.cpp [hunk 3]
```diff
@@ -278,6 +278,44 @@ constantFoldBinaryWithOverflow(BuiltinInst *BI, BuiltinValueKind ID,
            ResultsInError);
 }
 
+static SILValue countZeros(BuiltinInst *BI, llvm::Intrinsic::ID ID) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/ConstantFolding.cpp]: I find this name means nothing and this needs a comment. My suggestion:

```
/// Constant fold a cttz or ctlz builtin inst of an integer literal. If \p countLeadingZeros is set to true, then we
/// assume \p bi must be ctlz. If false, \p bi must be cttz.
///
/// NOTE: We assert that \p bi is either cttz or ctlz.
static SILValue constantFoldCountLeadingOrTrialingZeroIntrinsic(BuiltinInst *bi, bool countLeadingZeros) {
   ...
}
```

/// Count either the leading or trailing zeros.

Also, do you really need the intrinsic id here? Couldn't you just pass in a bool stating which of the two you are handling?

File: lib/SILOptimizer/Utils/ConstantFolding.cpp [hunk 4]
```diff
@@ -278,6 +278,44 @@ constantFoldBinaryWithOverflow(BuiltinInst *BI, BuiltinValueKind ID,
...
+  assert(BI->getArguments().size() == 2 && "Ctlz should have 2 args.");
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/ConstantFolding.cpp]: This is a bad assert message. You are handling both ctlz and cttz.

**Review by gottesmm — COMMENTED** [review]
Post commit review. Please fix.

**Comment by gottesmm** [comment]
Another thing. You can make all of this much simpler by using PatternMatch.h.

**Comment by shajrawi** [comment]
Thanks @gottesmm - https://github.com/apple/swift/pull/23739

**Pull request merged**